File: src/modules/equipment-model/services/equipment-model.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { EquipmentModel } from '../entities/equipment-model.entity';
import {
  CreateEquipmentModelDto,
  UpdateEquipmentModelDto,
} from '../dto/equipment-model.dto';
import { EquipmentCategory } from '../../equipment-category/entities/equipment-category.entity';

@Injectable()
export class EquipmentModelService {
  constructor(
    @InjectRepository(EquipmentModel)
    private readonly modelRepository: Repository<EquipmentModel>,
    @InjectRepository(EquipmentCategory)
    private readonly categoryRepository: Repository<EquipmentCategory>,
  ) {}

  findAll(): Promise<EquipmentModel[]> {
    return this.modelRepository.find({
      relations: { category: true },
      order: { name: 'ASC' },
    });
  }

  async findOne(id: string): Promise<EquipmentModel> {
    const model = await this.modelRepository.findOne({
      where: { id },
      relations: { category: true, assets: true },
    });
    if (!model) {
      throw new NotFoundException(`Equipment model with id "${id}" not found`);
    }
    return model;
  }

  async create(dto: CreateEquipmentModelDto): Promise<EquipmentModel> {
    const category = await this.categoryRepository.findOne({
      where: { id: dto.categoryId },
    });
    if (!category) {
      throw new NotFoundException('Category not found');
    }

    const model = this.modelRepository.create({
      name: dto.name,
      category,
      description: dto.description,
      defaultValue: dto.defaultValue ?? 0,
      procurementYear: dto.procurementYear,
      releaseYear: dto.releaseYear,
      expectedLifespanMonths: dto.expectedLifespanMonths,
      lowStockThreshold: dto.lowStockThreshold ?? 1,
    });

    return this.modelRepository.save(model);
  }

  async update(
    id: string,
    dto: UpdateEquipmentModelDto,
  ): Promise<EquipmentModel> {
    const model = await this.findOne(id);
    if (dto.categoryId) {
      const category = await this.categoryRepository.findOne({
        where: { id: dto.categoryId },
      });
      if (!category) {
        throw new NotFoundException('Category not found');
      }
      model.category = category;
    }
    Object.assign(model, {
      name: dto.name ?? model.name,
      description: dto.description ?? model.description,
      defaultValue: dto.defaultValue ?? model.defaultValue,
      procurementYear: dto.procurementYear ?? model.procurementYear,
      releaseYear: dto.releaseYear ?? model.releaseYear,
      expectedLifespanMonths:
        dto.expectedLifespanMonths ?? model.expectedLifespanMonths,
      lowStockThreshold: dto.lowStockThreshold ?? model.lowStockThreshold,
    });
    return this.modelRepository.save(model);
  }
}

File: src/EquipmentApi/Services/EquipmentModelService.cs
using EquipmentApi.Data;
using EquipmentApi.Dtos;
using EquipmentApi.Entities;
using EquipmentApi.Exceptions;
using Microsoft.EntityFrameworkCore;

namespace EquipmentApi.Services;

public class EquipmentModelService
{
    private readonly AppDbContext _db;

    public EquipmentModelService(AppDbContext db)
    {
        _db = db;
    }

    public Task<List<EquipmentModel>> FindAllAsync(CancellationToken cancellationToken = default)
    {
        return _db.EquipmentModels
            .Include(model => model.Category)
            .OrderBy(model => model.Name)
            .ToListAsync(cancellationToken);
    }

    public async Task<EquipmentModel> FindOneAsync(Guid id, CancellationToken cancellationToken = default)
    {
        var model = await _db.EquipmentModels
            .Include(m => m.Category)
            .Include(m => m.Assets)
            .FirstOrDefaultAsync(m => m.Id == id, cancellationToken);

        if (model is null)
        {
            throw new NotFoundException($"Equipment model with id \"{id}\" not found");
        }

        return model;
    }

    public async Task<EquipmentModel> CreateAsync(
        CreateEquipmentModelDto dto,
        CancellationToken cancellationToken = default)
    {
        var category = await _db.EquipmentCategories
            .FirstOrDefaultAsync(c => c.Id == dto.CategoryId, cancellationToken);
        if (category is null)
        {
            throw new NotFoundException("Category not found");
        }

        var model = new EquipmentModel
        {
            Name = dto.Name,
            CategoryId = category.Id,
            Description = dto.Description,
            DefaultValue = dto.DefaultValue ?? 0,
            ProcurementYear = dto.ProcurementYear,
            ReleaseYear = dto.ReleaseYear,
            ExpectedLifespanMonths = dto.ExpectedLifespanMonths,
            LowStockThreshold = dto.LowStockThreshold ?? 1,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow,
        };

        _db.EquipmentModels.Add(model);
        await _db.SaveChangesAsync(cancellationToken);
        model.Category = category;
        return model;
    }

    public async Task<EquipmentModel> UpdateAsync(
        Guid id,
        UpdateEquipmentModelDto dto,
        CancellationToken cancellationToken = default)
    {
        var model = await FindOneAsync(id, cancellationToken);

        if (dto.CategoryId is not null)
        {
            var category = await _db.EquipmentCategories
                .FirstOrDefaultAsync(c => c.Id == dto.CategoryId, cancellationToken);
            if (category is null)
            {
                throw new NotFoundException("Category not found");
            }

            model.CategoryId = category.Id;
            model.Category = category;
        }

        if (dto.Name is not null)
        {
            model.Name = dto.Name;
        }

        if (dto.Description is not null)
        {
            model.Description = dto.Description;
        }

        if (dto.DefaultValue is not null)
        {
            model.DefaultValue = dto.DefaultValue.Value;
        }

        if (dto.ProcurementYear is not null)
        {
            model.ProcurementYear = dto.ProcurementYear;
        }

        if (dto.ReleaseYear is not null)
        {
            model.ReleaseYear = dto.ReleaseYear;
        }

        if (dto.ExpectedLifespanMonths is not null)
        {
            model.ExpectedLifespanMonths = dto.ExpectedLifespanMonths;
        }

        if (dto.LowStockThreshold is not null)
        {
            model.LowStockThreshold = dto.LowStockThreshold.Value;
        }

        await _db.SaveChangesAsync(cancellationToken);
        return model;
    }
}
